File: 3329-find-the-length-of-the-longest-common-prefix/find-the-length-of-the-longest-common-prefix.cpp
class Solution {
public:
    int longestCommonPrefix(vector<int>& arr1, vector<int>& arr2) {
        unordered_map<string,int>mp;
        for(auto i:arr1)
        {
            string k=to_string(i);
            while(k!="")
            {
                mp[k]=0;
                k.pop_back();
            }
        }
        int res=0;
        for(auto i:arr2)
        {
           string k=to_string(i);
            while(res<k.size() and k!="" and mp.find(k)==mp.end())
            {
                k.pop_back();
            }
            if(mp.find(k)!=mp.end())
            {
                if(res<k.size())res=k.size();
            }
        }
        return res;
    }
};

File: 3329-find-the-length-of-the-longest-common-prefix/find-the-length-of-the-longest-common-prefix.java
class Solution {
    class TrieNode {
        TrieNode[] children = new TrieNode[10];
    }

    public int longestCommonPrefix(int[] arr1, int[] arr2) {
        TrieNode root = new TrieNode();
        int maxLen = 0;

        // 1. Build the Trie using arr2
        for (int num : arr2) {
            insert(num, root);
        }

        // 2. Search each number from arr1 in the Trie
        for (int num : arr1) {
            maxLen = Math.max(maxLen, findLongestPrefix(num, root));
        }

        return maxLen;
    }

    private void insert(int num, TrieNode root) {
        TrieNode curr = root;
        String s = Integer.toString(num);
        for (char c : s.toCharArray()) {
            int digit = c - '0';
            if (curr.children[digit] == null) {
                curr.children[digit] = new TrieNode();
            }
            curr = curr.children[digit];
        }
    }

    private int findLongestPrefix(int num, TrieNode root) {
        TrieNode curr = root;
        String s = Integer.toString(num);
        int count = 0;
        for (char c : s.toCharArray()) {
            int digit = c - '0';
            if (curr.children[digit] != null) {
                count++;
                curr = curr.children[digit];
            } else {
                break;
            }
        }
        return count;
    }
}